tensorflow/compiler/tf2tensorrt/trt_convert.cc

```diff
@@ -0,0 +1,389 @@
+                           ? "/job:localhost/replica:0/task:0/device:GPU:0"
+                           : node.device();
+  TF_RETURN_IF_ERROR(device_mgr->LookupDevice(device_name, &device));
+  *rm = device->resource_manager();
+  return Status::OK();
+}
+
+// Looks up the cache resurce associated with the TRT node.
+Status GetEngineCacheResource(const NodeDef& node, Session* session,
+                              TRTEngineCacheResource** resource) {
+  ResourceMgr* rm;
+  TF_RETURN_IF_ERROR(GetResourceManager(node, session, &rm));
+
+  absl::string_view resource_name = node.name();
+  size_t last_slash = resource_name.find_last_of('/');
+  if (last_slash != absl::string_view::npos) {
+    resource_name.remove_prefix(last_slash + 1);
+  }
+  const std::string container(kTfTrtContainerName);
+  *resource = nullptr;
+  TF_RETURN_IF_ERROR(
+      rm->Lookup(container, std::string(resource_name), resource));
+  if (resource == nullptr || (*resource)->cache_.size() == 0) {
+    return errors::Internal("Engine cache not found for", resource_name);
+  }
+  return Status::OK();
+}
+
+// Looks up the engine from the engine cache, and serializes the engine.
+Status ReadSerializedEngine(
+    const NodeDef& node, Session* session,
+    TrtUniquePtrType<nvinfer1::IHostMemory>* engine_data) {
+  TRTEngineCacheResource* resource;
+  TF_RETURN_IF_ERROR(GetEngineCacheResource(node, session, &resource));
+  core::ScopedUnref unref_cache_res(resource);
+  if (resource->cache_.size() > 1) {
+    return errors::Internal(
+        "Multiple engines found, but we can only serialize one");
+  }
+  const std::unique_ptr<EngineContext>& engine =
+      resource->cache_.begin()->second;
+  if (!engine || !engine->cuda_engine) {
+    // TODO(tfeher): Engine could have failed to build. Handle that correctly.
+    return errors::Internal("Engine not found for", node.name());
+  }
+
+  // Serialize the engine.
+  engine_data->reset(engine->cuda_engine->serialize());
+
+  return Status::OK();
+}
+
+// Saves the TRT engines a static attribute of the TRTEngineOp nodes.
+Status ConvertToStaticEngine(const GraphDef graph_def,
+                             GraphDef* static_graph_def, Session* session) {
+  static_graph_def->CopyFrom(graph_def);
+  VLOG(1) << "Saving TRT engines as static engine";
+  std::string op{"TRTEngineOp"};
+  for (auto& node : *(static_graph_def->mutable_node())) {
+    if (!op.compare(node.op())) {
+      VLOG(2) << "Saving TRT engine for " << node.name()
+              << ", device: " << node.device();
+      TrtUniquePtrType<nvinfer1::IHostMemory> engine_data;
+      TF_RETURN_IF_ERROR(ReadSerializedEngine(node, session, &engine_data));
+      auto* attr = node.mutable_attr();
+      AttrValue static_engine;
+      static_engine.set_b(true);
+      AttrValue engine_string;
+      engine_string.set_s(engine_data->data(), engine_data->size());
+      (*attr)["static_engine"] = static_engine;
+      (*attr)["serialized_segment"] = engine_string;
```

Comment:
Ideally, we should name these constants. But we can do it in another PR.

Reply:
Let us do it in the follow up PR.